### ISSUE TITLE
Posibility to execute same job in two different threads

### ISSUE THREAD

**Issue by troncomputers** [transistorsoft/flutter_background_fetch#248]
Hi.
I am not sure if I can ask about my "problem" here. I have very strange behavior with this lib. Is it possible to run same job in two different threads? 

I have one job that is going through documents and sends them to API. Sometimes I find duplicates in my ERP system, same document inserted twice, even if I have tons of conditions and one of them is checking if the document exists in database.
 
I did check API logs and:

```
2021-09-15 12:38:03.4947|0|INFO|PanelOptimo.Controllers.OptimoController|[INSERTING_USER] 
2021-09-15 12:38:03.4947|0|INFO|PanelOptimo.Controllers.OptimoController|[INSERTING_SERIE]
2021-09-15 12:38:03.4947|0|INFO|PanelOptimo.Controllers.OptimoController|[INSERTING_OBJECT]

2021-09-15 12:38:03.4947|0|INFO|PanelOptimo.Controllers.OptimoController|[INSERTING_USER]
2021-09-15 12:38:03.4947|0|INFO|PanelOptimo.Controllers.OptimoController|[INSERTING_SERIE]
2021-09-15 12:38:03.5108|0|INFO|PanelOptimo.Controllers.OptimoController|[INSERTING_OBJECT]
```
**[INSERTING_USER]** - which user account is inserting document
**[INSERTING_SERIE]** - which serie should document have
**[INSERTING_OBJECT]** - it is a document object as json string

Look at the times. How is that possible, that both `[INSERTING_USER]` log entries are at the same miliseconds? For me it looks like background_fetch is making the same API request at the same time (one task execute = two threads), On mobile device this document, for sure, is only once.

1. Is it happening because of this library?
2. Is it happening because of mobile device settings?

PS.
I was able to repeat this behavior only once, when I turn off internet connection and immediately turn it on again. I did try this method four more times and never happen again, but it is happening from time to time on my client devices.

**VERSION** 
`1.0.0-nullsafety.2`

**SETUP**
```dart
  MainScreen({Key? key}) : super(key: key) {
    initPlatformState();
    _onClickEnable(true);
    BackgroundFetch.registerHeadlessTask(backgroundFetchHeadlessTask);
  }

  Future<void> initPlatformState() async {
    int status = await BackgroundFetch.configure(
        BackgroundFetchConfig(
            minimumFetchInterval: 15,
            startOnBoot: true,
            stopOnTerminate: false,
            enableHeadless: true,
            forceAlarmManager: true,
            requiredNetworkType: NetworkType.ANY), (String taskId) async {
      print("[BackgroundFetch] Event received $taskId");
      print("TASK!: " + DateTime.now().toString());
      MainScreen.docController.documentSynchronization();
      BackgroundFetch.finish(taskId);
    }, (String taskId) async {
      print("[BackgroundFetch] TASK TIMEOUT taskId: $taskId");
      BackgroundFetch.finish(taskId);
    });

    BackgroundFetch.scheduleTask(TaskConfig(
        taskId: "com.foo.customtask",
        delay: 5000,
        forceAlarmManager: true,
        enableHeadless: true,
        requiredNetworkType: NetworkType.ANY,
        periodic: true));

    print('[BackgroundFetch] configure success: $status');
  }

  void _onClickEnable(enabled) {
    if (enabled) {
      BackgroundFetch.start().then((int status) {
        print('[BackgroundFetch] start success: $status');
      }).catchError((e) {
        print('[BackgroundFetch] start FAILURE: $e');
      });
    } else {
      BackgroundFetch.stop().then((int status) {
        print('[BackgroundFetch] stop success: $status');
      });
    }
  }
```

**Comment by troncomputers** [comment]
New info to this case

I used flutter_logs to save some data about numbers of tasks that are starting at the same time. I found it:
```
{MainScreen}  {initPlatformState}  {TASK_RUN = 2021-09-20 12:18:05.853110 ID = flutter_background_fetch}  {20 September 2021 12:18:05 PM}  {INFO}
{DocumentsController}  {documentSynchronization}  {DOCS_TO_SEND = 0}  {20 September 2021 12:18:05 PM}  {INFO}
{MainScreen}  {initPlatformState}  {TASK_RUN = 2021-09-20 12:18:05.909433 ID = com.foo.customtask}  {20 September 2021 12:18:05 PM}  {INFO}
{DocumentsController}  {documentSynchronization}  {DOCS_TO_SEND = 0}  {20 September 2021 12:18:05 PM}  {INFO}
```
My configured task has id `com.foo.customtask` but there is another one named `flutter_background_fetch`. Is my initialization created incorrectly? I am not able to find anywhere in my code text `flutter_background_fetch`

**Comment by christocracy** [comment]
`fetchtask` is the traditional task you should expect to see fired according to configured `minimumFetchInterval`.

.scheduleTask is a relatively recent add-on to the plugin.  It is not required, and on iOS fires much less frequently.

You should rely on fetchtask.

**Comment by troncomputers** [comment]
It is mostly "copy/paste". Would you be able to write me here how should my setup look? I want to fire `MainScreen.docController.documentSynchronization();` only once every 15 minutes.

**Comment by christocracy** [comment]
Delete call to `.scheduleTask`.

**Comment by christocracy** [comment]
In practice, your fetch task won't be executed *every* 15 minutes.  It totally depends on the user's behavior with your app.  If the user doesn't engage frequently with your app, the OS will execute your task less frequently.

If the user doesn't open your app daily, iOS won't execute tasks at all, as noted in the readme.

**Comment by troncomputers** [comment]
I've noticed that. My software is running only on Android. Those 15 minutes are the most optimal to run synchronization, but if it depends on usage and is going to run more frequently it doesn't matter. For now the most important thing is to stop making duplicates (run twice). I'm going to make changes that you suggested and if it's going to work I will close this issue ✌

**Comment by christocracy** [comment]
> `BackgroundFetch.registerHeadlessTask(backgroundFetchHeadlessTask);`

You are not supposed to do this for Android in your Flutter App instance.  It's supposed to placed in `main.dart`, exactly as documented.

**Comment by troncomputers** [comment]
I can't do that right now. I need to be logged in to synchronize documents. The easiest (fastest) way was to create separate `MainPage` that opens up after `LoginPage`. This `MainPage` was necessary for other parts of the application and I did not create it just for schedule. I will fix that to ;-)

**Comment by christocracy** [comment]
Your Android headlessTask won't even work the way you have it configured.  It will likely crash your app.

**Comment by troncomputers** [comment]
I moved all configuration to `main()`. Created another task to check if user is logged in and all required data is synchronized from API to cache. If `loggedIn == true` and `synchronizeDate != null` then I'm creating main task with documents synchronizaton. How to stop first one if all of them have the same `taskId = flutter_background_fetch`?

**Comment by christocracy** [comment]
See [Debugging](https://github.com/transistorsoft/flutter_background_fetch#android-1) section of README.  Learn to *simulate* events and observe `logcat` logs.

**Comment by troncomputers** [comment]
It is still not working properly. I am still having problem with executing twice at the same time.

**EXAMPLE LOGS**
```
{MainScreen}  {initPlatformState}  {TASK_RUN = 2021-09-28 11:21:48.361609 ID = flutter_background_fetch}  {28 September 2021 11:21:48 AM}  {INFO}
{DocumentsController}  {documentSynchronization}  {DOCS_TO_SEND = 1}  {28 September 2021 11:21:48 AM}  {INFO}
{MainScreen}  {initPlatformState}  {TASK_RUN = 2021-09-28 11:21:48.423227 ID = flutter_background_fetch}  {28 September 2021 11:21:48 AM}  {INFO}
{DocumentsController}  {documentSynchronization}  {DOCS_TO_SEND = 1}  {28 September 2021 11:21:48 AM}  {INFO}

{MainScreen}  {initPlatformState}  {TASK_RUN = 2021-09-28 11:36:47.541696 ID = flutter_background_fetch}  {28 September 2021 11:36:47 AM}  {INFO}
{DocumentsController}  {documentSynchronization}  {DOCS_TO_SEND = 1}  {28 September 2021 11:36:47 AM}  {INFO}
{MainScreen}  {initPlatformState}  {TASK_RUN = 2021-09-28 11:36:47.598542 ID = flutter_background_fetch}  {28 September 2021 11:36:47 AM}  {INFO}
{DocumentsController}  {documentSynchronization}  {DOCS_TO_SEND = 1}  {28 September 2021 11:36:47 AM}  {INFO}

{MainScreen}  {initPlatformState}  {TASK_RUN = 2021-09-28 11:51:48.219492 ID = flutter_background_fetch}  {28 September 2021 11:51:48 AM}  {INFO}
{DocumentsController}  {documentSynchronization}  {DOCS_TO_SEND = 0}  {28 September 2021 11:51:48 AM}  {INFO}
{MainScreen}  {initPlatformState}  {TASK_RUN = 2021-09-28 11:51:48.234638 ID = flutter_background_fetch}  {28 September 2021 11:51:48 AM}  {INFO}
{DocumentsController}  {documentSynchronization}  {DOCS_TO_SEND = 0}  {28 September 2021 11:51:48 AM}  {INFO}
```

Entry named `TASK_RUN` has more information (contains miliseconds). How is that still possible?

**Comment by christocracy** [comment]
Are you running your app with live-update, causing `BackgroundFetch.configure` to run multiple times?

I'm not interested in your own custom logs.  I'm only interested in logs from:
```
$ adb logcat *:S TSBackgroundFetch:V
```

**Comment by troncomputers** [comment]
It's not easy to get logcat. I was able to do it remotely. Is that enough?
```
--------- beginning of main
09-29 09:31:25.509 27962 27962 D TSBackgroundFetch: - Background Fetch event received: flutter_background_fetch
09-29 09:31:25.514 27962 27962 D TSBackgroundFetch: ­čĺÇ [HeadlessTask flutter_background_fetch]
09-29 09:31:25.518 27962 27962 D TSBackgroundFetch: - Background Fetch event received: flutter_background_fetch
09-29 09:31:25.521 27962 27962 D TSBackgroundFetch: ­čĺÇ [HeadlessTask flutter_background_fetch]
09-29 09:31:25.522 27962 27962 D TSBackgroundFetch: - finish: flutter_background_fetch
09-29 09:31:25.523 27962 27962 D TSBackgroundFetch: - FetchAlarmReceiver finish
09-29 09:31:25.529 27962 27962 D TSBackgroundFetch: - finish: flutter_background_fetch
09-29 09:31:25.530 27962 27962 D TSBackgroundFetch: - FetchAlarmReceiver finish
09-29 10:01:25.969 27962 27962 D TSBackgroundFetch: - Background Fetch event received: flutter_background_fetch
09-29 10:01:25.972 27962 27962 D TSBackgroundFetch: ­čĺÇ [HeadlessTask flutter_background_fetch]
09-29 10:01:25.975 27962 27962 D TSBackgroundFetch: - finish: flutter_background_fetch
09-29 10:01:25.976 27962 27962 D TSBackgroundFetch: - FetchAlarmReceiver finish
09-29 10:01:25.981 27962 27962 D TSBackgroundFetch: - Background Fetch event received: flutter_background_fetch
09-29 10:01:25.984 27962 27962 D TSBackgroundFetch: ­čĺÇ [HeadlessTask flutter_background_fetch]
09-29 10:01:25.987 27962 27962 D TSBackgroundFetch: - finish: flutter_background_fetch
09-29 10:01:25.988 27962 27962 D TSBackgroundFetch: - FetchAlarmReceiver finish
09-29 10:31:26.271 27962 27962 D TSBackgroundFetch: - Background Fetch event received: flutter_background_fetch
09-29 10:31:26.273 27962 27962 D TSBackgroundFetch: ­čĺÇ [HeadlessTask flutter_background_fetch]
09-29 10:31:26.278 27962 27962 D TSBackgroundFetch: - Background Fetch event received: flutter_background_fetch
09-29 10:31:26.281 27962 27962 D TSBackgroundFetch: ­čĺÇ [HeadlessTask flutter_background_fetch]
09-29 10:31:26.284 27962 27962 D TSBackgroundFetch: - finish: flutter_background_fetch
09-29 10:31:26.286 27962 27962 D TSBackgroundFetch: - FetchAlarmReceiver finish
09-29 10:31:26.289 27962 27962 D TSBackgroundFetch: - finish: flutter_background_fetch
09-29 10:31:26.291 27962 27962 D TSBackgroundFetch: - FetchAlarmReceiver finish
09-29 12:31:26.685 27962 27962 D TSBackgroundFetch: - Background Fetch event received: flutter_background_fetch
09-29 12:31:26.702 27962 27962 D TSBackgroundFetch: ­čĺÇ [HeadlessTask flutter_background_fetch]
09-29 12:31:26.708 27962 27962 D TSBackgroundFetch: - finish: flutter_background_fetch
09-29 12:31:26.709 27962 27962 D TSBackgroundFetch: - FetchAlarmReceiver finish
09-29 12:31:26.712 27962 27962 D TSBackgroundFetch: - Background Fetch event received: flutter_background_fetch
09-29 12:31:26.716 27962 27962 D TSBackgroundFetch: ­čĺÇ [HeadlessTask flutter_background_fetch]
09-29 12:31:26.722 27962 27962 D TSBackgroundFetch: - finish: flutter_background_fetch
09-29 12:31:26.723 27962 27962 D TSBackgroundFetch: - FetchAlarmReceiver finish
09-29 14:31:22.363 27962 27962 D TSBackgroundFetch: - Background Fetch event received: flutter_background_fetch
09-29 14:31:22.374 27962 27962 D TSBackgroundFetch: ­čĺÇ [HeadlessTask flutter_background_fetch]
09-29 14:31:22.388 27962 27962 D TSBackgroundFetch: - Background Fetch event received: flutter_background_fetch
09-29 14:31:22.391 27962 27962 D TSBackgroundFetch: ­čĺÇ [HeadlessTask flutter_background_fetch]
09-29 14:31:22.436 27962 27962 D TSBackgroundFetch: - finish: flutter_background_fetch
09-29 14:31:22.437 27962 27962 D TSBackgroundFetch: - FetchAlarmReceiver finish
09-29 14:31:22.439 27962 27962 D TSBackgroundFetch: - finish: flutter_background_fetch
09-29 14:31:22.440 27962 27962 D TSBackgroundFetch: - FetchAlarmReceiver finish
--------- beginning of system
09-29 15:58:12.811 27962 27962 D TSBackgroundFetch: - Background Fetch event received: flutter_background_fetch
09-29 15:58:12.822 27962 27962 D TSBackgroundFetch: ­čĺÇ [HeadlessTask flutter_background_fetch]
09-29 15:58:12.846 27962 27962 D TSBackgroundFetch: - Background Fetch event received: flutter_background_fetch
09-29 15:58:12.851 27962 27962 D TSBackgroundFetch: ­čĺÇ [HeadlessTask flutter_background_fetch]
09-29 15:58:12.922 27962 27962 D TSBackgroundFetch: - finish: flutter_background_fetch
09-29 15:58:12.923 27962 27962 D TSBackgroundFetch: - FetchAlarmReceiver finish
09-29 15:58:12.927 27962 27962 D TSBackgroundFetch: - finish: flutter_background_fetch
09-29 15:58:12.929 27962 27962 D TSBackgroundFetch: - FetchAlarmReceiver finish

```

**Comment by christocracy** [comment]
Looking at the `/example` app on two devices *Pixel 3a @ 12* and *Samsung Galaxy S20 @ 11*.
- *Pixel 3a*:  no double events.
- *Samsung*:  periodic double events.

I suspect the OS is queuing events while in *Doze Mode* and firing the queued event(s) in the next processing cycle.

While I look into this (time consuming to reproduce), I suggest you record a timestamp `lastFetchEventAt` when each event occurs.  Compare the current timestamp with `lastFetchEventAt`.  Based upon the time-difference (eg: `< 1000ms`), choose to simply call `BackgroundFetch.finish(taskId)` and `return` without processing.
